WcaOnRails/app/webpacker/components/CompetitionOverview/CompetitionFilters.js

```diff
@@ -0,0 +1,662 @@
+import React, { useState, useEffect } from 'react';
+import { useInView } from 'react-intersection-observer';
+import { useInfiniteQuery } from '@tanstack/react-query';
+import {
+  Button, Icon, Form, Container, Dropdown, Popup, List, Input, Header,
+} from 'semantic-ui-react';
+import DatePicker from 'react-datepicker';
+
+import I18n from '../../lib/i18n';
+import { competitionsApiUrl, delegatesApiUrl } from '../../lib/requests/routes.js.erb';
+import { fetchJsonOrError } from '../../lib/requests/fetchWithAuthenticityToken';
+import {
+  events, continents, countries, competitionConstants,
+} from '../../lib/wca-data.js.erb';
+
+import CompetitionTable from './CompetitionTable';
+import CompetitionMap from './CompetitionMap';
+
+import 'react-datepicker/dist/react-datepicker.css';
+
+// Max number of competitions fetched per query
+const COMPETITIONS_API_PAGINATION = 25;
+
+// Limit number of markers on map, especially for "All Past Competitions"
+const MAP_DISPLAY_LIMIT = 500;
+
+const WCA_EVENT_IDS = Object.values(events.official).map((e) => e.id);
+
+const PAST_YEARS_WITH_COMPETITIONS = [];
+for (let year = new Date().getFullYear(); year >= 2003; year -= 1) {
+  PAST_YEARS_WITH_COMPETITIONS.push(year);
+}
+PAST_YEARS_WITH_COMPETITIONS.push(1982);
+
+const regionsOptions = [
+  { key: 'all', text: I18n.t('common.all_regions'), value: 'all' },
+  {
+    key: 'continents_header', value: '', disabled: true, content: <Header content={I18n.t('common.continent')} size="small" style={{ textAlign: 'center' }} />,
+  },
+  ...(Object.values(continents.real).map((continent) => (
+    { key: continent.id, text: continent.name, value: continent.id }
+  ))),
+  {
+    key: 'countries_header', value: '', disabled: true, content: <Header content={I18n.t('common.country')} size="small" style={{ textAlign: 'center' }} />,
+  },
+  ...(Object.values(countries.real).map((country) => (
+    { key: country.id, text: country.name, value: country.iso2 }
+  ))),
+];
+
+function CompetitionFilters() {
+  const [competitionApiKey, setCompetitionApiKey] = useState({
+    sort_by: 'present', year: '', region: '', delegate: '', search: '',
+  });
+  const [selectedEvents, setSelectedEvents] = useState([]);
+  const [selectedRegion, setSelectedRegion] = useState();
+  const [selectedDelegate, setSelectedDelegate] = useState();
+  const [pastSelectedYear, setPastSelectedYear] = useState('all_years');
+  const [customStartDate, setCustomStartDate] = useState();
+  const [customEndDate, setCustomEndDate] = useState();
```

Comment:
Due to being so new to React I wasn't even aware `useReducer` was a thing... it is indeed so much better than using this many states. What's even better is that the state generated by `useReducer` can be used as the query key straight away.

Reply:
Given that it's called `filterState` it might make sense to include _all_ these filter-based states in it (that is, selected events, should show cancelled, etc.). The other benefit of doing that is that right now the reducer vs individual states are split between those which are handled by the back-end vs the front-end which means that if one of those gets handled differently later then the reducer would have to be modified. If everything's in the reducer from the start then nothing would have to change.

That last point is a bit hypothetical though and it's not always a good use of time to prepare for things that might not happen.

Reply:
Now that I'm no longer using the entire `filterState` object as the query key I agree this should be done. It would previously have been a problem since changing the selected events would generate a new query key whereas filtering by events is not done by the API. 

I talked with Gregor about filtering in the front-end vs the back-end and I think eventually we will move everything to the back-end (not necessarily in this PR, and maybe except showing cancelled competitions?).

Reply:
Should `filterReducer`, `filterInitialState`, and `calculateQueryKey` be separated into another file?